gematria/granite/python/token_graph_builder_model.py

```diff
@@ -388,52 +377,64 @@ class TokenGraphBuilderModelNodeEmbed(snt.Embed):
 
   def __init__(
       self,
+      common_embed_dim,
+      num_annotations,
       instruction_annotations,
       instruction_node_mask,
       **kwargs,
   ) -> None:
     """Initializes node embeddings.
 
     Args:
+      common_embed_dim: The length of the learnt part of the instruction node
+        embedding vectors. The remainder of the vector is filled with
+        instruction annotation.
+      num_annotations: The number of annotations per instruction.
       instruction_annotations: Tensor holding instruction level runtime
         annotations as in `BasicBlockGraphBuilder`.
       instruction_node_mask: As in `BasicBlockGraphBuilder`.
-      **kwargs: Additional arguments to be passed to the internal `snt.Embed`.
+      kwargs: Additional arguments to be passed to the internal `snt.Embed`s.
     """
-    super().__init__(**kwargs)
-
-    # The number of annotations per instruction.
-    self.num_annotations = int(instruction_annotations.shape[1])
-
-    if self.num_annotations > self.embed_dim:
-      raise ValueError('num_annotations cannot be greater than embed_dim.')
-
     self.instruction_annotations = instruction_annotations
     self.instruction_node_mask = instruction_node_mask
 
-  def __call__(self, inputs):
-    embeddings = super().__call__(inputs)
+    # The first `embed_dim - num_annotations` embedding values for all nodes.
+    self.common_embed = snt.Embed(
```

Comment:
We should probably mark all the attributes as internal (by using a leading underscore). I don't see a need for a user to access those.